Sources/Bond/Collections/ObservableArray.swift

```diff
@@ -295,30 +296,30 @@ extension MutableObservableArray {
 }
 
 extension MutableObservableArray where Item: Equatable {
-  
+
   public func replace(with array: [Item], performDiff: Bool) {
-    if performDiff {
-      lock.lock()
+    lock.lock(); defer { lock.unlock() }
 
+    if performDiff {
       let diff = self.array.extendedDiff(array)
+      let patch = diff.patch(from: self.array, to: array)
```

Comment:
Thanks for confirming you hunch! :)

I've been playing with this in the morning and got an idea of an operator on the signal of observable array events that would, basically, do other kind of batching. 

In short, we intercept `beginBatchEditing` and at that point save the array in a local variable. Then we ignore all events (default case) until we get `endBatchEditing`. At that point we calculate new diff and patch and send that out.

I have not tested this well as I have to leave now, but give it a try :) Just call this operator before binding to an NSTableView.

```swift
public extension SignalProtocol where Element: ObservableArrayEventProtocol, Element.Item: Equatable {

  public func patchedBatch() -> Signal<ObservableArrayEvent<Item>, Error> {
    var isBatching = false
    var originalArray: [Item] = []
    return Signal { observer in
      return self.observe { event in
        switch event {
        case .next(let observableArrayEvent):
          switch observableArrayEvent.change {
          case .beginBatchEditing:
            isBatching = true
            originalArray = observableArrayEvent.source.array
            observer.next(ObservableArrayEvent(change: .beginBatchEditing, source: observableArrayEvent.source))
          case .endBatchEditing:
            isBatching = false
            let array = observableArrayEvent.source.array
            let diff = originalArray.extendedDiff(array)
            let patch = diff.patch(from: originalArray, to: array)
            for step in patch {
              switch step {
              case .insertion(let index, _):
                observer.next(ObservableArrayEvent(change: .inserts([index]), source: observableArrayEvent.source))
              case .deletion(let index):
                observer.next(ObservableArrayEvent(change: .deletes([index]), source: observableArrayEvent.source))
              case .move(let from, let to):
                observer.next(ObservableArrayEvent(change: .move(from, to), source: observableArrayEvent.source))
              }
            }
            observer.next(ObservableArrayEvent(change: .endBatchEditing, source: observableArrayEvent.source))
          default:
            if !isBatching {
              observer.next(ObservableArrayEvent(change: observableArrayEvent.change, source: observableArrayEvent.source))
            }
          }
        case .failed(let error):
          observer.failed(error)
        case .completed:
          observer.completed()
        }
      }
    }
  }
}
```